### ISSUE TITLE
False positive for boxed local in default Trait method implementation

### ISSUE THREAD

**Issue by MentalAllergen** [rust-lang/rust-clippy#4804]
Output of `cargo clippy -V`: `clippy 0.0.212 (3aea8603 2019-09-03)`

Given the following code:
```rust
trait Foo {
    fn bar(self: Box<Self>) -> u32 {
        5
    }
}
```

Clippy lint [boxed local](https://rust-lang.github.io/rust-clippy/master/index.html#boxed_local) is triggered, advising us to remove the Box and have plain `self` in the argument list, like this:
```rust
trait Foo {
    fn bar(self) -> u32 {
        5
    }
}
```

However, doing this makes the code fail compilation with the following error: 
`the size for values of type "Self" cannot be known at compilation time`, 
which is always the case for traits.

I've encountered this while implementing the state design pattern with dynamic dispatch. I wouldn't usually opt for dynamic dispatch, however the problem domain requires it.

Ideally this lint should check weather the implementation is a default trait method implementation before triggering.

**Comment by MentalAllergen** [comment]
Looking at the implementation, I should be able to make a fix for this. Will give it a try and report here. How would I test the changes locally?

**Comment by flip1995** [comment]
Hey @MentalAllergen, sorry for the late reply. I think every Clippy maintainer is currently pretty busy.

How the testing in Clippy work is described here: https://github.com/rust-lang/rust-clippy/blob/master/doc/adding_lints.md#testing

TL;DR: Add a test case for this in the corresponding test file in `tests/ui/` and then test it with `TESTNAME=boxed_local cargo +master uitest` (I'm not sure if the corresponding test file is really called `boxed_local.rs`). You can install the master toolchain with the `setup-toolchain.sh` script in the project root.

**Comment by MentalAllergen** [comment]
Thanks for the explanation @flip1995!
PR is up.

I still couldn't test locally as `cargo +master build` failed for the `master` branch of clippy. It seems to be a problem with a change from `rust-lang` itself so until that is resolved, testing will not be possible.

**Comment by benmkw** [comment]
I think this is another/ the same case where this lint may not be accurate https://github.com/probe-rs/probe-rs/pull/437/files#file-probe-rs-src-probe-mod-rs-L459. I tried to apply the suggestion at some point but I'm pretty convinced that its not possible.
